man/man8/zfs.8

```diff
@@ -983,6 +1009,35 @@ This property can also be referred to by its shortened column name,
 
 .sp
 .ne 2
+.na
+\fB\fBencryption\fR=\fBoff\fR | \fBon\fR | \fBaes-128-ccm\fR | \fBaes-192-ccm\fR | \fBaes-256-ccm\fR | \fBaes-128-gcm\fR | \fBaes-192-gcm\fR | \fBaes-256-gcm\fR
+.ad
+.sp .6
+.RS 4n
+Controls the encryption cipher suite (block cipher, key length, and mode) used for this dataset. Requires the encryption feature to be enabled on the pool. Requires a keysource to be set at dataset creation time.
+.sp
+Selecting \fBencryption\fR=\fBon\fR when creating a dataset indicates that the current default encryption suite will be selected, which is currently \fBaes-256-ccm\fR. In order to provide consistent data protection, encryption must be specified at dataset creation time and it cannot be changed afterwards.
+.sp
+The \fBencryption\fR property is inherited by child datasets by default. All inheriting datasets will be accessible when the root key is mounted. A child dataset may also set encryption locally, which will indicate that it should have a separate key and \fBkeysource\fR from the parent. Regardless, all children of an encrypted dataset must also be encrypted.
+.sp
+For more details and caveats about encryption see \fBzpool-features\fR.
+.RE
+
+.sp
+.ne 2
+.mk
+.na
+\fB\fBkeysource\fR=<\fBraw\fR | \fBhex\fR | \fBpassphrase\fR>,<\fBprompt\fR | \fBfile://\fR\fI<absolute file path\fR>>
+.ad
+.sp .6
+.RS 4n
+Controls what format the user's encryption key will be provided as and where it should be loaded from. This property is only set when the dataset is encrypted.
+.sp
+Raw keys and hex keys must be 32 bytes long (regardless of the chosen encryption suite) and must be randomly generated. Passphrases must be between 8 and 64 bytes long and will be processed through PBKDF2 before being used (see the \fBpbkdf2iters\fR property). Even though the encryption suite cannot be changed after dataset creation, the keysource can be with \fBzfs key -c\fR. If prompt is selected as the key location ZFS will ask for the key at the command prompt when it is required (\fBzfs mount\fR and \fBzfs key -l\fR). If \fBfile\fR is selected, the key will be loaded from the specified absolute file path. The key may also be passed in via STDIN, but users should be careful not to place keys which should be kept secret on the command line.
```

Comment:
> So if we can zfs set keysource= to change the location of the key, when exactly do we re-encrypt the master keys?

We only have to re-encrypt the master keys when we actually change the bits of the wrapping key (more pbkdf2 iterations, new pbkdf2 salt or changing the key itself). This requires that the existing key is loaded since we need the master key decrypted in the first place. Currently, however, it is implemented such that even changing the keysource requires the keys to be loaded.

> It seems like we want to allow changing the location of the key when it is not loaded, so we have to be able to change it without re-encrypting the master key. It seems like we would want the same behavior w.r.t. re-encrypting whether the keys are loaded or not. Therefore zfs set keysource= should not re-encrypt the master key. So maybe we say that only zfs key -c changes the wrapping key (as it works today) and re-encrypts the master key. Then zfs key -c -o keysource= is equivalent to zfs set keysource=value; zfs key -c

So in that case, I think I would argue that we should remove `zfs key -c -o keysource=` as a capability, replacing it with `zfs set`. This way `zfs key -c` manages all of the changes that require the keys to be loaded and zfs set manages everything else.

> Incidentally, could this (zfs key -c) be used to change the passphrase? Presumably it would prompt for the passphrase? (In which case it can also reference the STDIN stuff in zfs key -l.)

Yes, in fact that is probably its primary use case. Right now it requires that you do the `zfs key -l` manually beforehand. I'm not sure if the 2 steps should be combined because it leads to some interesting questions from a UI perspective. For instance, if the keys are unloaded when I run `zfs key -c`, should I expect them to be loaded or unloaded once the command has returned?